### DISCUSSION TITLE
Resources handling in GDevelop

### DISCUSSION THREAD

**Discussion by AlexandreSi** [4ian/GDevelop#4032] [Enhancement ideas (and their technical discussions)]
We can see a few suggestions/issues around resource handling (for instance the cleaning process).
This discussion is open to gather any feedback around it.

**Comment by AlexandreSi** [comment]
About resource/object removal: https://github.com/4ian/GDevelop/issues/3014

**Comment by Entr0py404** [comment]
I made up this as a prototype in paint, Could moving resources work like this?
the combobox would contain all directories and sub directories in the project folder
![Screenshot 2022-06-13 235102](https://user-images.githubusercontent.com/75917656/173517034-c790be12-07c2-4332-bc05-725f434204c7.png)

**Reply by AlexandreSi** [reply]
With the `assets\` prefix on the resources list, I guess you want to give a notion of hierarchy?
I'm having trouble understanding the left part: you type in the folder where you would like the 4 selected resources to go? And create the folder on the fly?

**Reply by AlexandreSi** [reply]
> With the assets\ prefix on the resources list, I guess you want to give a notion of hierarchy?

Oups, just realized the `assets\` part is already implemented

**Reply by arthuro555** [reply]
He is showing how he is selecting multiple resources at once and that displays a different resource details screen: instead of showing the resource's settings, it shows a field that allows to batch move them to another folder. This would likely completely move the files there too, not just change the path the resources point to.

**Comment by arthuro555** [comment]
- When automatically moving an image into the project folder or creating an asset with piskel, resources should be copied to the assets folder instead of the root project folder, like the asset store files
- Big projects often have a tree of folders to sort and find assets quickly. GDevelop resource fields allow you to either find a resource by the resource name or create a new resource. Instead, the create a resource option should be named pick from filesystem, and not recreate a resource when picking a file that already has a resource associated to it.
- If a resource is not named after the file it links to, scanning for resources or adding one from the file system will create a new one. GDevelop should check if a resource already is using this file instead of checking if one is named after that file.
- All places that allow using a resource should propose getting one from the asset store. For example, selecting a resource for a sprite frame or picking a music resource in events should always do so, not just the web app.
- Resource default names for files that are in the folder "assets" shouldn't contain the "assets/" in their resource name, as ideally all resources are in here and it doesn't help differentiating resources from one another.
- Resources have settings, like preloading audio files, but they are pretty hidden away from the average user inside the resource tab. When picking a file that didn't have a resource associated yet, GDevelop shouldn't silently create the resource but allow directly modifying those settings; Either by directly opening a "new resource settings" dialogue, or by having a small dismissible hint popping up that says "resource created" with a button to open those settings (so that you are not bothered by that dialogue when adding multiple resources to your project at once, for example by adding many frames to an animation of a sprite). I'd go with the second one personally as I think it'd be too intrusive and inconvenient to power users otherwise.
- Custom resources should be available: they'd be resources you can create in the resource tab and can be either a binary or text file. You would then be able to get their contents as text if it is a text file or as url if it is a binary one through expressions. That is useful for loading in data in a custom text format like PBS files for pokemon fangames, or binary files like a zip file containing DLC data.

**Comment by VegeTato** [comment]
Well arthuro already covered everything :p
but there is one thing I think worth to mention
in the latest GDevelop patch it says:
> Added an option to remove all unused resources

that's cool, but if possible can we have the option (**Scan all resources**) as well ?
currently the process is:
**• Scan Audio** > **Scan Image** > **Scan Font** > **Scan Video** > **Scan Json** > **Scan Bitmap font**
**• Remove unused Audio** > **Remove unused Image** > **Remove unused Font** > **Remove unused Video** > **Remove unused Json** > **Remove unused Bitmap font**

After the latest update it is:
**• Scan Audio** > **Scan Image** > **Scan Font** > **Scan Video** > **Scan Json** > **Scan Bitmap font**
**• Remove All unused Resources**

How am suggesting to make it:
**Scan all resources** > **Remove All unused Resources**

2 clicks can do the job :3

**Reply by arthuro555** [reply]
@Oxey405 You want to try to do this?

**Reply by Oxey405** [reply]
Sure I could try :)

**Comment by tristanbob** [comment]
Thanks for looking into improvements on the Resource page!  I like what others have said so far.

One thing that has always bothered me is that there are some actions that apply to all resources, but to initiate them you have to right-click on a single resource?  This includes:

- Scan in the project for...
- Remove unused...
- Remove resources with invalid path

Instead, those "global" functions should be buttons that are not tied to a specific object.  These buttons could be placed in the top right (near the existing buttons), but they don't have to go there.
